src/condor_credd/condor_credmon_oauth/condor_credmon_oauth

```diff
@@ -80,22 +94,68 @@ def start_credmon_process(q, cred_dir, args):
     credmon_incomplete(cred_dir)
     logger = setup_logging(**vars(args))
 
-    credmons = [
-        OAuthCredmon(cred_dir, args),
-        LocalCredmon(cred_dir, args)
-    ]
+    # Each provider needs to "stay in their lane" and be told which service providers to use.
+    # - Administrators must spell out the provider configuration for the OAuth2 and Local credmons.
+    # - Users can specify arbitrary provider names and configurations for the Vault credmon
+    # For backward compat, if the providers aren't explicitly set, we use the following algorithm:
+    # - If it's `scitokens`, we assume it is a local credmon.
+    # - If config knob `FOO_CLIENT_ID` is non-empty, then we assume it's the OAuth2 credmon.
+    # - Otherwise, it's the Vault credmon's.
+
+    # The local credmon historically had a single provider; the Vault and OAuth credmon assumed they provided all tokens.
+    local_provider_names = set(filter(None, re.compile(r"[\s,]+").split(htcondor.param.get("LOCAL_CREDMON_PROVIDER_NAMES", ""))))
+    # Backward compat: if not providing a list, fallback to the singular config value.
+    if not local_provider_names:
+        local_provider_names = {htcondor.param.get("LOCAL_CREDMON_PROVIDER_NAME", "scitokens")}
+    oauth2_providers = set(filter(None, re.compile(r"[\s,]+").split(htcondor.param.get("OAUTH2_CREDMON_PROVIDER_NAMES", "*"))))
+    client_providers = set(filter(None, re.compile(r"[\s,]+").split(htcondor.param.get("CLIENT_CREDMON_PROVIDER_NAMES", ""))))
+
+    credmons = \
+        [OAuthCredmon(providers=oauth2_providers, cred_dir=cred_dir, args=args)] + \
+        [LocalCredmon(provider=provider, cred_dir=cred_dir, args=args) for provider in local_provider_names] + \
+        [ClientCredmon(provider=provider, cred_dir=cred_dir, args=args) for provider in client_providers]
+
+    vault_providers = set(filter(None, re.compile(r"[\s,]+").split(htcondor.param.get("VAULT_CREDMON_PROVIDER_NAMES", "*"))))
+    if VaultCredmon is None and vault_providers and vault_providers != {"*"}:
+        logger.warning("Vault credmon is configured via `VAULT_CREDMON_PROVIDER_NAMES` but the Vault credmon is not installed")
+        vault_providers = set()
+    elif VaultCredmon is not None:
+        credmons.append(VaultCredmon(providers=vault_providers, cred_dir=cred_dir, args=args))
+
+    # Ugly n^2 check for overlapping providers
+    providers = {
+        "local": local_provider_names,
+        "oauth2": oauth2_providers,
+        "client": client_providers,
+        "vault": vault_providers,
```

Comment:
Should these all end, or all not end, in `_names`?  AFAICT, all of them are the same thing.

Reply:
Cleaned up in 866d438789662139287bdceda6d41ae81414f8f4